src/lib/core/cord_on_demand.cc

```diff
@@ -0,0 +1,53 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2010-2022, Tarantool AUTHORS, please see AUTHORS file.
+ */
+
+#include "cord_on_demand.h"
+
+#include <stdlib.h>
+
+#include "fiber.h"
+#include "trivia/util.h"
+
+/**
+ * RAII wrapper around a cord object.
+ *
+ * A thread-local cord object is created on the first call to the get() method
+ * and destroyed when the thread exits.
+ */
+class CordOnDemand final {
+public:
+	static cord *get()
+	{
+		static thread_local CordOnDemand singleton;
+		return singleton.cord_ptr;
+	}
+
+private:
+	struct cord *cord_ptr;
```

Comment:
Nit: I think this is overengineering — this doesn't provide any encapsulation, since one can still get this pointer through `get`.

Better off making this member public and changing the `get` method to:
```C++
static CordOnDemand &instance() noexcept {
	thread_local CordOnDemand singleton;
	return singleton;
}
```

Reply:
> I think this is overengineering — this doesn't provide any encapsulation, since one can still get this pointer through `get`.

But one can't change the value of the `cord_ptr` pointer with `get` while with `instance` one can. Besides `CordOnDemand()::get()` looks better than `CordOnDemand()::instance().cord_ptr` IMO.

Reply:
Okay, sounds reasonable.

Reply:
Let’s also make the `get` method const-qualified then.

Reply:
> Besides CordOnDemand()::get() looks better than CordOnDemand()::instance().cord_ptr.

Perhaps a matter of taste, but `_ptr` suffix looks ugly to me.

Reply:
> Let’s also make the `get` method const-qualified then.

[A static member function cannot have a cv-qualifier](https://stackoverflow.com/questions/19812786/why-cant-a-static-member-function-have-a-cv-qualifier).



> Perhaps a matter of taste, but `_ptr` suffix looks ugly to me.

Without the suffix, it wouldn't compile:

```
/home/vlad/src/tarantool/tarantool/src/lib/core/cord_on_demand.cc:29:15: error: declaration of ‘cord* CordOnDemand::cord’ changes meaning of ‘cord’ [-fpermissive]
   29 |  struct cord *cord;
      |               ^~~~
In file included from /home/vlad/src/tarantool/tarantool/src/lib/core/cord_on_demand.cc:11:
/home/vlad/src/tarantool/tarantool/src/lib/core/fiber.h:707:8: note: ‘cord’ declared here as ‘struct cord’
  707 | struct cord {
      |        ^~~~
```

We could use `_` suffix, as they usually add to private members, but we don't follow this rule in other parts of our code. Besides the `cord_ptr` name is consistent with [`cord_ptr`](https://github.com/tarantool/tarantool/blob/256da01046da8f4fbe81d629cf54ae32dc04a1fb/src/lib/core/fiber.h#L784) defined in the fiber module.